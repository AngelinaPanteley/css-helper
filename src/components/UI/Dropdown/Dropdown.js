import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import styles from './Dropdown.scss';
import Backdrop from '../Backdrop/Backdrop';
import Auxiliary from '../../../hoc/Auxiliary/Auxiliary';

class Dropdown extends PureComponent {
  static propTypes = {
    isOpen: PropTypes.bool.isRequired,
    onClick: PropTypes.func,
    onClose: PropTypes.func.isRequired,
    left: PropTypes.bool,
    right: PropTypes.bool,
    links: PropTypes.array,
  }

  onLinkToClick = () => {
    this.props.onClose();

    if (this.props.onClick) {
      this.props.onClick();
    }
  }

  onLinkClick = (e, link) => {
    e.preventDefault();
    this.onLinkToClick();
    link.click();
  }

  render() {
    const attachedClasses = [styles.Dropdown];

    if (this.props.isOpen) {
      attachedClasses.push(styles.Open);
    } else {
      attachedClasses.push(styles.Close);
    }

    if (this.props.left) {
      attachedClasses.push(styles.Left);
    } else if (this.props.right) {
      attachedClasses.push(styles.Right);
    }

    return (
      <Auxiliary>
        <Backdrop show={this.props.isOpen} clicked={this.props.onClose} />
        <div className={attachedClasses.join(' ')}>
          <ul className={styles.List}>
            {
              this.props.links && this.props.links.map((link) => {
                return (
                  <li key={link.title} className={styles.List_Item}>
                    {
                      link.route
                        ?
                        <Link to={link.route}
                          onClick={this.onLinkToClick}>
                          {link.title}
                        </Link>
                        :
                        <a onClick={(e) => this.onLinkClick(e, link)}>
                          {link.title}
                        </a>
                    }
                  </li>
                )
              })
            }
          </ul>
        </div>
      </Auxiliary>
    );
  }
};

export default Dropdown;