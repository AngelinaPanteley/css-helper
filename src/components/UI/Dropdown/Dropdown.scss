@import '../../../variables';

.Dropdown {
  position: absolute;
  width: auto;
  max-width: 70%;
  height: calc(100vh - #{$header-height});
  top: $header-height;
  z-index: map-get($z-index, popup);
  background-color: $color-gray;
  box-sizing: border-box;
  transition: transform 0.3s ease-out;

  &.Open {
    transform: translateX(0);
  }

  &.Close {
    transform: translateX(-100%);

    &.Right {
      transform: translateX(100%);
    }
  }

  &.Left {
    left: 0;
  }

  &.Right {
    right: 0;
  }
}

.List {
  padding: 0;
  margin: 0;
  list-style: none;
}

.List_Item {
  margin: 20px 0;

  a {
    padding: 10px 43px;
    background: $color-dark-pink;
    width: 100%;
    display: inline-block;
    text-decoration: none;

    &:hover,
    &:active,
    &:focus {
      color: white;
      background: $color-light-pink;
    }
  }
}