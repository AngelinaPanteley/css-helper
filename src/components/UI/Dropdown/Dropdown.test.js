"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import { expect as chaiExpect } from 'chai';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16.1';
import Dropdown from './Dropdown.js';
import { MemoryRouter, Link } from "react-router-dom";
configure({ adapter: new Adapter() });

describe('>>>Dropdown', () => {

  //SNAPSHOTS

  const props = {
    isOpen: true,
    onClick: () => { },
    onClose: () => { },
    left: true,
    right: false,
    links: [{
      title: 'title',
      route: '/route',
    }],
  }

  const dropdownComponent = <MemoryRouter><Dropdown {...props} /></MemoryRouter>;

  const component = renderer.create(dropdownComponent);

  it('renders itself correctly', () => {
    let componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();
  });

  //SHALLOW RENDERING

  let wrapper;
  const shallowComponent = <Dropdown {...props} />;

  beforeEach(() => {
    wrapper = shallow(shallowComponent);
  });

  it('renders list of links', () => {
    chaiExpect(wrapper.find('ul')).to.have.lengthOf(1);
  });

})