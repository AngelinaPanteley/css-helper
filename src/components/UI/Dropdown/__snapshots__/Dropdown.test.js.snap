// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>Dropdown renders itself correctly 1`] = `
Array [
  <div
    className=""
    onClick={[Function]}
  />,
  <div
    className="  "
  >
    <ul>
      <li>
        <a
          href="/route"
          onClick={[Function]}
        >
          title
        </a>
      </li>
    </ul>
  </div>,
]
`;
